model/hex_bits_signed.sail

```diff
@@ -0,0 +1,141 @@
+/*  This software was developed by SRI International and the University of  */
+/*  Cambridge Computer Laboratory (Department of Computer Science and       */
+/*  Technology) under DARPA/AFRL contracts FA8650-18-C-7809 ("CIFV")        */
+/*  and FA8750-10-C-0237 ("CTSRD").                                         */
+/*                                                                          */
+/*  Redistribution and use in source and binary forms, with or without      */
+/*  modification, are permitted provided that the following conditions      */
+/*  are met:                                                                */
+/*  1. Redistributions of source code must retain the above copyright       */
+/*     notice, this list of conditions and the following disclaimer.        */
+/*  2. Redistributions in binary form must reproduce the above copyright    */
+/*     notice, this list of conditions and the following disclaimer in      */
+/*     the documentation and/or other materials provided with the           */
+/*     distribution.                                                        */
+/*                                                                          */
+/*  THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS''      */
+/*  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED       */
+/*  TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A         */
+/*  PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR     */
+/*  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,            */
+/*  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT        */
+/*  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF        */
+/*  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND     */
+/*  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,      */
+/*  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT      */
+/*  OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF      */
+/*  SUCH DAMAGE.                                                            */
+/*==========================================================================*/
+
+$ifndef _HEX_BITS_SIGNED
+$define _HEX_BITS_SIGNED
+
+$include <vector.sail>
+$include <string.sail>
+
+val parse_hex_bits_signed : forall 'n, 'n > 0. (int('n), string) -> bits('n)
+val valid_hex_bits_signed : forall 'n, 'n > 0. (int('n), string) -> bool
+
+val hex_bits_signed: forall 'n, 'n > 0. bits('n) <-> (int('n), string)
+
+function hex_bits_signed_forwards(bv) = {
+  if signed(bv) < 0
+  then (length(bv), concat_str("-", hex_str(unsigned(not_vec(bv) + 1))))
+  else (length(bv), hex_str(unsigned(bv)))
+}
+function hex_bits_signed_forwards_matches(bv) = true
+
+function parse_hex_bits_signed(n, str) = {
+  if string_take(str, 1) == "-"
+  then {
+    let str_ = string_drop(str, 1);
```

Comment:
Btw you don't actually need a different name here. You can just do `let str = string_drop(str, 1);` and it essentially makes a new variable that shadows the old one. I generally prefer it in situations like this since it prevents you from accidentally using the old one in later code.